SLING-12714 - Impementation of OidcAuthenticationHandler (Support for PKCE, nonce in OidcAuthenticationHandler, bug fixes)

nscendoni opened the pull request:
- Added nonce validation
- Added support for Authorization Code Flow with PKCE
- Refactoring of IT Tests
- Improved README.md
- Added multiple paths in OidcAuthenticationHandler configuration
- Added redirect to original url in OidcAuthenticationHandler
- Renamed UserInfoProcessot to SlingUserInfoProcessor
- Added configuration to store access and refresh with encryption to SlingUserInfoProcessor

rombert reviewed (commented):
75 commits seems a bit excessive, can you please split into logical commits?

nscendoni commented:
> 75 commits seems a bit excessive, can you please split into logical commits?

Yes I did it. I didn't realized that I pushed from an old branch we already merged.

anchela commented:
@nscendoni , i edited the title to make sure it refers to an existing jira ticket.

anchela commented on src/main/java/org/apache/sling/auth/oauth_client/impl/RedirectHelper.java:
```diff
@@ -16,57 +16,135 @@
     // consent, account selection, 2FA, etc. so we cannot make this too short.
     private static final int COOKIE_MAX_AGE_SECONDS = 300;
-    
+    static Logger logger = LoggerFactory.getLogger(RedirectHelper.class);
```
imho this logger should be private... will check in an ide

anchela replied:
done

anchela commented on src/main/java/org/apache/sling/auth/oauth_client/impl/RedirectHelper.java:
```diff
@@ -16,57 +16,135 @@
         return cookie;
     }
+
+    public static String findLongestPathMatching(@NotNull String[] path,@Nullable String url) {
```
at a first glance i didn't see this utility used outside of RedirectHelper -> make it private.

the return value can be null, right? -> annotate accordingly.

anchela replied:
i was mistaken -> resolving

anchela commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -282,25 +304,61 @@ public AuthenticationInfo extractCredentials(@NotNull HttpServletRequest request
+            ClientID clientId = new ClientID(conn.clientId());
+            AuthorizationCode code = new AuthorizationCode(authCode);
+
+            TokenRequest tokenRequest;
```
this new code is quite hard to read.... i wonder if we can simplify it.

anchela reviewed (approved):
i left a few comments, mostly nitpicking. and will try to fix it right away.... i would love them to be addressed but not worth rejecting the PR.

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -120,7 +126,7 @@ public class OidcAuthenticationHandler extends DefaultAuthenticationFeedbackHand
                 description = "Repository path for which this authentication handler should be used by Sling. If this is " +
                         "empty, the authentication handler will be disabled. By default this is set to \"/\".")
-        String path() default "/";
+        String[] path() default {"/"};
```
Should we rename to paths?

nscendoni replied:
Actually it is from: https://github.com/apache/sling-org-apache-sling-auth-core/blob/47eee11b0a05bdd3d20fcbfba81b2077f503f058/src/main/java/org/apache/sling/auth/core/spi/AuthenticationHandler.java#L44-L57

rombert replied:
Make sense, thanks.

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -159,11 +168,13 @@ public OidcAuthenticationHandler(@Reference(policyOption = ReferencePolicyOption
         this.defaultConnectionName = config.defaultConnectionName();
         this.userInfoProcessor = userInfoProcessor;
         this.userInfoEnabled = config.userInfoEnabled();
+        this.checkNonce = config.checkNonce();
```
What is this in reference to? I browsed https://datatracker.ietf.org/doc/html/rfc6749 but don't see anything special (except for a non-normative example in https://datatracker.ietf.org/doc/html/rfc6749#section-7.1 ).

rombert replied:
Oh, this is the OIDC handler, right. Quoting from the spec

>  If present in the ID Token, Clients MUST verify that the nonce Claim Value is equal to the value of the nonce parameter sent in the Authentication Request.

What is the reason for making this configurable?

nscendoni replied:
the nonce parameter is not mandatory in the request, but if it is present the IdP must send it in the ID Token.
I made it configurable in case some IdP do not support it.

rombert replied:
So you are concerned about sending a nonce in the authentication request to an authorization server that does not support nonces?

nscendoni replied:
Yes. If the server would not support it, we would not validate the token.

rombert replied:
I realise I'm being pedantic, but I don't like seeing config options that should not be there :-) You don't have to remove if you're really sure that there are going to be non-compliant OIDC server implementations out there we need to support.

Otherwise I would always keep it enabled because it enhances security and disabling it would bring no benefit.

nscendoni replied:
I removed that configuration.

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -187,32 +198,40 @@ public AuthenticationInfo extractCredentials(@NotNull HttpServletRequest request
         Optional<OAuthState> clientState; //state returned by the idp in the redirect request
         String authCode; //authorization code returned by the idp in the redirect request
         Cookie stateCookie;
+        Cookie nonceCookie;
```
I would prefer to use a single cookie, if at all possible. This makes the response payload smaller.

nscendoni replied:
@rombert @anchela 
Currently in oauth flow we send several information in the state:
- perRequestKey (random value to be verified)
- connectionName
- redirect
The perRequestKey is also sent in one encrypted cookie to be verified for CSRF protection.
During OIDC authentication flow, additional information (nonce and code verifier) are sent into separated cookies.

If you agree we may send a single json encrypted cookie with all the information required by both the flows, and send only the perRequestKey in the state.

nscendoni replied:
In [rfc6265](https://datatracker.ietf.org/doc/html/rfc6265?#section-6.1) read that these are the limits that browsers SHOULD support:
```
   o  At least 4096 bytes per cookie (as measured by the sum of the
      length of the cookie's name, value, and attributes).
   o  At least 50 cookies per domain.
   o  At least 3000 cookies total.
```
If we want to be in a safer side and limit the number of cookies as well we may:
- keep the current setup for OAuth
- Modify the current code to have only one new cookie for nonce and code verifier

rombert replied:
If we are sure we can fit within 4096 byes I would suggest that we keep one cookie, this should make the payload smaller and the code easier to understand.

I am not sure if it makes sense to have different coookie names for the OAuth servlets vs the Authentication Handler, to simplify the code. Just an idea, not a suggestion :-)

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -187,32 +198,40 @@ public AuthenticationInfo extractCredentials(@NotNull HttpServletRequest request
-        } catch (ParseException | URISyntaxException e) {
-            logger.debug("Failed to parse authorization response");
+        } catch (ParseException | URISyntaxException | IllegalStateException e) {
+            // If we fail parsing the response, we consider the request not for this authentication handler
```
Did you remove the debug logging on purpose? I think it could be useful for troubleshooting

rombert commented on src/test/resources/keycloak-import/sling.json:
```diff
@@ -351,7 +352,7 @@
   "otpPolicyPeriod" : 30,
   "otpPolicyCodeReusable" : false,
-  "otpSupportedApplications" : [ "totpAppGoogleName", "totpAppFreeOTPName" ],
+  "otpSupportedApplications" : [ "totpAppFreeOTPName", "totpAppGoogleName" ],
```
Is this change needed? If not, please revert it.

rombert commented on src/test/resources/keycloak-import/sling.json:
```diff
@@ -1237,7 +1285,7 @@
   "supportedLocales" : [ ],
   "authenticationFlows" : [ {
-    "id" : "bd07af7f-436a-4c7d-a2ae-78e20a897cb3",
+    "id" : "aff611f8-41d9-49ad-ba2b-7a24ae030bc8",
```
I think this (and the following ID changes) are not necessary. It would be better to exclude them so we don't pollute the history.

nscendoni replied:
I did several changes on the exported keycloak configuration:

1. Added a new client that support pkce authentication
2. Modified the user profile to have few attributes
3. Defined few groups, added the test user as member and added them to the claim returned to the clients

rombert reviewed (changes requested):
Thanks for the PR. Some very high-level comments from me ( and the assorted nitpicks )

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/ClientConnection.java:
```diff
@@ -34,4 +35,10 @@ public interface ClientConnection {
      * @return the name of the connection
      */
     @NotNull String name();
+    String clientId();
```
@nscendoni - the `ClientConnection` was designed to not leak implementation details to the client. What was the reason to add all these field to it?

BTW, it does not seem related to the 'move to Java 11' commit :-)

rombert commented:
@nscendoni - I think it would be useful to separate 3d4330474a55b236270a30560ed37722094ed36e into a different Jira issue and pull request. Agreed we should downgrade to Java 11 (for now) to target a wider deployment base.

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/ClientConnection.java:
```diff
@@ -34,4 +36,10 @@ public interface ClientConnection {
      * @return the name of the connection
      */
     @NotNull String name();
+    @NotNull String clientId();
```
Same comment as in https://github.com/apache/sling-org-apache-sling-auth-oauth-client/pull/17#discussion_r2099767014 , please don't expose this kind of information in `ClientConnection`

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -159,11 +168,13 @@ public OidcAuthenticationHandler(@Reference(policyOption = ReferencePolicyOption
         this.stateManager = stateManager;
         this.idp = config.idp();
         this.callbackUri = config.callbackUri();
-        this.defaultRedirect = config.defaultRedirect();
```
Why did you remove this config setting?

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/impl/SlingUserInfoProcessorImpl.java:
```diff
@@ -0,0 +1,154 @@
+
+@Component(
+        service = UserInfoProcessor.class,
+        immediate = true,
```
Do you need immediate here?  It will be activated as soon as a reference is requested. You maybe also want to set the configuration policy to required.

nscendoni replied:
I removed it

rombert commented on src/main/java/org/apache/sling/auth/oauth_client/spi/UserInfoProcessor.java:
```diff
@@ -17,15 +17,13 @@
  */
 public interface UserInfoProcessor {
-    @NotNull OidcAuthCredentials process(@Nullable UserInfo userInfo, @NotNull TokenResponse tokenResponse, 
+    @NotNull OidcAuthCredentials process(@Nullable String userInfo, @NotNull String tokenResponse,
```
I think it would be good to document the String values here; links to the RFC would IMO be enough.

rombert reviewed (changes requested):
I think this is very close to being done. From my point of view there are only three issues to solve:

- the cookie topic (important)
- removal of unrelated formatting changes, e.g. adding empty lines, cleaning up whitespace,etc. This makes history much harder to track. Any formatting changes must  come as a separate PR and we should do it by upgrading to the latest parent pom which enforces a code formatting standard. But let's do that after we merge this PR :-)
- the ITs leave untracked files in the project's base dir, please review the persistence code. Also the file name seems wrong (no extension)


```
	cookie-tokens14644194543422998595bin
	cookie-tokens17544281903593273246bin
	cookie-tokens401491827215654835bin
	cookie-tokens5223146744397278699bin
	cookie-tokens5292745495627831483bin
	cookie-tokens5503793274250268725bin
	cookie-tokens6559516317127871037bin
```

nscendoni commented:
> I think this is very close to being done. From my point of view there are only three issues to solve:
> 
> * the cookie topic (important)

I've pushed a change that works with only one encrypted cookie. I also refactored the OAuth Code to work with that.
There are not anymore information passed in the state parameter.

> * removal of unrelated formatting changes, e.g. adding empty lines, cleaning up whitespace,etc. This makes history much harder to track. Any formatting changes must  come as a separate PR and we should do it by upgrading to the latest parent pom which enforces a code formatting standard. But let's do that after we merge this PR :-)

I fixed the setting in my ide  :-)

> * the ITs leave untracked files in the project's base dir, please review the persistence code. Also the file name seems wrong (no extension)
> 
> ```
> 	cookie-tokens14644194543422998595bin
> 	cookie-tokens17544281903593273246bin
> 	cookie-tokens401491827215654835bin
> 	cookie-tokens5223146744397278699bin
> 	cookie-tokens5292745495627831483bin
> 	cookie-tokens5503793274250268725bin
> 	cookie-tokens6559516317127871037bin
> ```

I fixed it.
